File: LRU Cache.java
public class LRUCache {
    class LinkedList {
        LinkedList(int k, int v) {
        	key = k;
            value = v;
            prev = null;
            next = null;
        }
        int key;
        int value;
        LinkedList prev;
        LinkedList next;
    }
    private int m_capacity;
    private Map<Integer, LinkedList> m_cache;
    LinkedList m_head;
    LinkedList m_tail;
    public LRUCache(int capacity) {
        m_capacity = capacity;
        m_cache = new HashMap<Integer, LinkedList>();
        m_head = null;
        m_tail = null;
    }

    public int get(int key) {
        if (m_cache.containsKey(key)) {
            touch(key);
            return m_head.value;
        }
        return -1;
    }

    public void set(int key, int value) {
    	if (!m_cache.containsKey(key) && m_cache.size() == m_capacity) {
            // remove least used key
            LinkedList tail = m_tail.prev;
            m_cache.remove(m_tail.key);
            m_tail = tail;
        }
        touch(key);
        m_cache.get(key).value = value;
    }

    private void touch(int key) {
        if (!m_cache.containsKey(key)) {
            LinkedList node = new LinkedList(key, -1);
            m_cache.put(key, node);
            if (m_head == null) {
            	m_head = node;
            	m_tail = node;
            	return;
            }
            node.next = m_head;
            m_head.prev = node;
            m_head = node;
            return;
        }
        LinkedList node = m_cache.get(key);
        if (node == m_head)
        	return;
        if (node == m_tail) {
        	m_tail = node.prev;
        	m_tail.next = null;
        	node.prev = null;
        	node.next = m_head;
        	m_head.prev = node;
        	m_head = node;
        	return;
        }
    	// remove node
        if (node.prev != null)
            node.prev.next = node.next;
        if (node.next != null)
            node.next.prev = node.prev;

        // put node to the head
        node.prev = null;
        node.next = m_head;
        m_head.prev = node;
        m_head = node;
    }
}

File: LRU Cache.py
class LRUCache:

	# @param capacity, an integer
	def __init__(self, capacity):
		self.cache = {}
		self.capacity = capacity
		self.order = []

	def touch(self, key):
		if (key in self.order):
			self.order.remove(key)
		self.order.insert(0, key)

	# @return an integer
	def get(self, key):
		if (not key in self.order):
			return -1
		self.touch(key)
		return self.cache[self.order[0]]

	# @param key, an integer
	# @param value, an integer
	# @return nothing
	def set(self, key, value):
		if (not key in self.cache):
			if (len(self.cache) == self.capacity):			
				del self.cache[self.order.pop()]
		self.touch(key)
		self.cache[key] = value
